syntax = "proto3";
package kv;

message KeyValue {
  string key = 1;
  string value = 2;
  int32 version = 3;
}

message ReplicateRequest {
  KeyValue kv = 1;
  string from_node = 2;
}

message ReplicateResponse {
  bool ok = 1;
  string message = 2;
}

message StreamBucketRequest {
  int32 bucket_id = 1;
  int32 start = 2;
  int32 limit = 3;
}

message StreamKV {
  KeyValue kv = 1;
}

service Replicator {
  rpc Replicate(ReplicateRequest) returns (ReplicateResponse);
  // Server-side streaming for bucket transfer
  rpc StreamBucket(StreamBucketRequest) returns (stream StreamKV);
}
